File: thereminimum.cpp
#include <iostream>
#include <cstring>
#include "Leap.h"

using namespace Leap;

class SampleListener : public Listener {
  public:
    virtual void onInit(const Controller&);
    virtual void onConnect(const Controller&);
    virtual void onDisconnect(const Controller&);
    virtual void onExit(const Controller&);
    virtual void onFrame(const Controller&);

  private:
};

const std::string fingerNames[] = {"Thumb", "Index", "Middle", "Ring", "Pinky"};
const std::string boneNames[] = {"Metacarpal", "Proximal", "Middle", "Distal"};
const std::string stateNames[] = {"STATE_INVALID", "STATE_START", "STATE_UPDATE", "STATE_END"};

void SampleListener::onInit(const Controller& controller) {
  std::cout << "Initialized" << std::endl;
}

void SampleListener::onConnect(const Controller& controller) {
  std::cout << "Connected" << std::endl;
  controller.enableGesture(Gesture::TYPE_CIRCLE);
  controller.enableGesture(Gesture::TYPE_KEY_TAP);
  controller.enableGesture(Gesture::TYPE_SCREEN_TAP);
  controller.enableGesture(Gesture::TYPE_SWIPE);
}

void SampleListener::onDisconnect(const Controller& controller) {
  // Note: not dispatched when running in a debugger.
  std::cout << "Disconnected" << std::endl;
}

void SampleListener::onExit(const Controller& controller) {
  std::cout << "Exited" << std::endl;
}

void SampleListener::onFrame(const Controller& controller) {
  // Get the most recent frame and report some basic information
  
  const Frame frame = controller.frame();
  /*
  std::cout << "Frame id: " << frame.id()
            << ", timestamp: " << frame.timestamp()
            << ", hands: " << frame.hands().count()
            << ", extended fingers: " << frame.fingers().extended().count()
            << ", tools: " << frame.tools().count()
            << ", gestures: " << frame.gestures().count() << std::endl;
  */
  HandList hands = frame.hands();

  if (hands.count()>2)
  {
    std::cout << "More than two hands. Nope." << std::endl;
    exit(123);
  } 

  if (hands.count()==1)
  {
    Hand hand = hands[0];
    const double pitch = 5.*hand.palmPosition()[1];
    const double roll = (hand.isLeft())? hand.palmNormal().roll() *RAD_TO_DEG : -1.*hand.palmNormal().roll() *RAD_TO_DEG;
    
    double vol = (roll + 90)/180;
    vol = ( vol > 1) ? 1 : vol ;
    vol = ( vol < 0) ? 0 : vol ;

    if (!frame.hands().isEmpty()){
      std::cout << "(" << pitch << ", " << vol << ")" << std::endl;
    }
  }
  else if (hands.count()==2)
  {
    Hand handL, handR;
    if ((hands[0].isLeft() && hands[1].isLeft()) || (hands[0].isRight() && hands[1].isRight()))
    {
      std::cout << "Both hands of the same sign. Nope." << std::endl;
      exit(123);
    } else if (hands[0].isLeft() && hands[1].isRight())
    {
      handL=hands[0];
      handR=hands[1];
    } else if (hands[1].isLeft() && hands[0].isRight())
    {
      handR=hands[0];
      handL=hands[1];
    }
  
    const double pitchL = 5.*handL.palmPosition()[1];
    const double rollL = handL.palmNormal().roll() *RAD_TO_DEG ;
    
    double volL = (rollL + 90)/180;
    volL = ( volL > 1) ? 1 : volL ;
    volL = ( volL < 0) ? 0 : volL ;
  
    const double pitchR = 5.*handR.palmPosition()[1];
    const double rollR = -1.*handR.palmNormal().roll() *RAD_TO_DEG;
    
    double volR = (rollR + 90)/180;
    volR = ( volR > 1) ? 1 : volR ;
    volR = ( volR < 0) ? 0 : volR ;
  
  
    if (!hands.isEmpty()){
        std::cout << "(" << pitchL << ", " << volL << ", " << pitchR << ", " << volR << ")" << std::endl;
    }
  }

}
int main(int argc, char** argv) {

  Controller controller;
  SampleListener listener;

  controller.addListener(listener);

  if (argc > 1 && strcmp(argv[1], "--bg") == 0)
    controller.setPolicy(Leap::Controller::POLICY_BACKGROUND_FRAMES);
  
  
  std::cout << "Press Enter to quit..." << std::endl;
  std::cin.get();

  controller.removeListener(listener);
  
  return 0;
}

File: thereminimum.py
# -*- coding: utf-8 -*-
"""thereminimum.py

Written on 20/01 by Will Grant, Tomé Gouveia and Brett Abram

Takes input from Leap motion and writes it out to stdout.

It writes out the palm position coordinates as:
(x, y, z)
"""

import sys, thread, time, os, pyaudio, threading, Queue

if sys.platform == 'linux2':
    sys.path.append('LeapSDK/lib/x64')
    import LeapPython
sys.path.append('LeapSDK/lib')
import Leap

from Leap import CircleGesture, KeyTapGesture, ScreenTapGesture, SwipeGesture

import palmPositionToSound as ppts


class SampleListener(Leap.Listener):
    def on_init(self, controller):
        print "Initialized"

    def on_connect(self, controller):
        print "Connected"

    def on_disconnect(self, controller):
        # Note: not dispatched when running in a debugger.
        print "Disconnected"

    def on_exit(self, controller):
        print "Exited"

    def on_frame(self, controller):
        # Get the most recent frame and report some basic information
        frame = controller.frame()
        hand = frame.hands[0]
        # If there's a hand write out formatted palm positions to screen.
        if not frame.hands.is_empty:
            # Get the hand's normal vector and direction
            yaw = hand.palm_normal.roll if hand.is_left else -1. * hand.palm_normal.roll
            vol = (yaw * Leap.RAD_TO_DEG + 90) / 180
            if vol > 1:
                vol = 1
            if vol < 0:
                vol = 0

            pitch = 15 * frame.hands[0].stabilized_palm_position[1]
            # pitch = 15 * frame.hands[0].palm_position[1]

            q.put(ppts.LeapData(pitch, vol))


# queue for passing messages between threads
q = Queue.Queue()


def main():
    # Create a sample listener and controller
    listener = SampleListener()
    controller = Leap.Controller()

    # start thread
    t = threading.Thread(target=ppts.play_sound, args=(q, ))
    # daemon won't stop program from exiting when it's the only thread left
    t.daemon = True
    t.start()

    # Have the sample listener receive events from the controller
    controller.add_listener(listener)

    # Keep this process running until Enter is pressed
    print "Press Enter to quit..."
    try:
        sys.stdin.readline()
    except KeyboardInterrupt:
        pass
    finally:
        # Remove the sample listener when done
        controller.remove_listener(listener)


if __name__ == "__main__":
    main()
